### PR TITLE
Sign up for Cargorilla from the lobby

### PR DESCRIPTION
**Pull request by Jacquerel**
## About The Pull Request

If the station rolls the "Cargo Gorilla" trait, a button will now be visible on the lobby.
Clicking on this button before the round has started will add you to a list of participants, one of whom will be selected to become a gorilla when the round begins.
If nobody signs up (because they're really boring I guess) the job will instead appear on the latejoin menu.
Once someone has become the gorilla the button will disappear.

dreamseeker_ntP3OayAuV

While implementing this I noticed that an inverted check means we were never populating the "GLOB.cargo_sloth" field which means the station trait wasn't even working.

BEHIND THE SCENES
This also adds a generic "job station trait" which can be expanded in the future.
Future developers can extend this to add other "rare jobs" with relative ease.
By default I have made it so all subtypes of this trait are mutually exclusive, only one can roll at a time.

This also means that I have converted "cargo gorilla" into a job, which applies most of the code previously located in the mob's typepath or in the station trait.
The fact that it is a job means that **admins** can enable any number of gorillas to be present on the latejoin menu (but not the roundstart one, as it is not possible to add Cargo Gorilla to your occupation preferences) if they so desire.
The random beurocratic station trait, event, and traitor item (and the job console) are not able to add gorilla slots.

Because I changed "Cargo Gorilla" to a job it now no longer exists on the map until a player gains the role, and there wasn't a non-hacky way to copy the name of this round's cargo sloth. Instead I just added a small cargo gorilla name list.

## Why It's Good For The Game

Makes the presence of a fun trait more visible to players.
Means that people who aren't observing get a chance to be a monkey.
This is a framework several other people have wanted to exist for their own features.

## Changelog

:cl: Jacquerel and Fikou
qol: If the station rolls the "Cargo Gorilla" station trait. you will be able to sign up for the role from the game lobby.
qol: If nobody signs up to be the Cargo Gorilla then you can select it from the Late Join menu and arrive on the arrival shuttle.
fix: The Cargo Gorilla will actually spawn.
/:cl:

### CODE REVIEW

**Comment by MrMelbert** [comment]
Kinda sad that it ruins the Selective Attention Test

**Comment by Jacquerel** [comment]
> Kinda sad that it ruins the Selective Attention Test

You might still fail to see the gorilla on the lobby screen but it is much less likely

**Comment by optimumtact** [comment]
image
soon

**Comment by CCC23** [comment]
gorrilia button

**Comment by Mothblocks** [comment]
why is the tooltip so far below the button? also i think that only one of these should be active at a time to begin with so

**Comment by Jacquerel** [comment]
> why is the tooltip so far below the button? also i think that only one of these should be active at a time to begin with so

If anyone could tell me the answer to that I would appreciate it, I do not know
I'll dig into what it is doing some time tomorrow though probably

**Review by Wallemations — APPROVED** [review]
Gorilla button

File: code/datums/station_traits/neutral_traits.dm [hunk 1]
```diff
@@ -145,14 +145,67 @@
...
 	. = ..()
 	RegisterSignal(SSatoms, COMSIG_SUBSYSTEM_POST_INITIALIZE, PROC_REF(replace_cargo))
+	RegisterSignal(SSdcs, COMSIG_GLOB_PRE_GAMEMODE_SETUP, PROC_REF(on_gamemode_setup))
+
+/datum/station_trait/cargorilla/can_display_lobby_button()
+	return accepting_signups
+
+/datum/station_trait/cargorilla/setup_lobby_button(atom/movable/screen/lobby/button/sign_up/lobby_button)
+	RegisterSignal(lobby_button, COMSIG_ATOM_UPDATE_OVERLAYS, PROC_REF(on_lobby_button_update_overlays))
+	lobby_button.desc = "Become the Cargo Gorilla, a peaceful shepherd of boxes."
```
Comment (Fikou) [inline code/datums/station_traits/neutral_traits.dm:168]:
```suggestion
	lobby_button.desc = "Sign up for Cargo Gorilla, a peaceful shepherd of boxes."
```
people arent guaranteed to get it

Comment (BlueMemesauce) [reply code/datums/station_traits/neutral_traits.dm:168]: They are if its latejoin

Comment (Fikou) [reply code/datums/station_traits/neutral_traits.dm:168]: now were at an impasse

Comment (Mothblocks) [reply code/datums/station_traits/neutral_traits.dm:168]: It shouldn't be on the lobby when the round has started, it should show on the normal job selection menu.

Comment (Jacquerel) [reply code/datums/station_traits/neutral_traits.dm:168]: OK I can remove it after the round has started (it did this initially but then I thought "but what if...")

**Comment by Jacquerel** [comment]
I still havent figured out why the tooltip is over there unfortunately but it does now only show during roundstart (as default but overridable behaviour for any trait with an associated button).

File: code/_onclick/hud/new_player.dm [hunk 2]
```diff
@@ -6,32 +6,53 @@
...
+
+/// Display buttons for relevant station traits
+/datum/hud/new_player/proc/add_station_trait_buttons()
+	if (!mymob || !mymob.client || mymob.client.interviewee || !length(GLOB.lobby_station_traits))
+		return
+	var/position_iterator = 1
+	for (var/datum/station_trait/trait as anything in GLOB.lobby_station_traits)
+		if (!trait.can_display_lobby_button())
+			continue
+		var/atom/movable/screen/lobby/button/sign_up/sign_up_button = new(null, /* our_hud = */ src)
```
Comment (Ghommie) [inline code/_onclick/hud/new_player.dm:40]:
```suggestion
		var/atom/movable/screen/lobby/button/sign_up/sign_up_button = new(our_hud = src)
```
That's how it was done on L21.

File: code/datums/station_traits/neutral_traits.dm [hunk 3]
```diff
@@ -181,19 +216,56 @@
...
-		return
+/// Called before we start assigning roles
+/datum/station_trait/cargorilla/proc/on_gamemode_setup()
+	SIGNAL_HANDLER
+	if (!LAZYLEN(lobby_candidates))
+		return // Nobody signed up :(
+	if (QDELETED(cargorilla) || cargorilla.key || cargorilla.stat == DEAD)
+		return // Something weird happened to our buddy Cargorilla
+
+	for (var/mob/dead/signee in lobby_candidates)
```
Comment (Ghommie) [inline code/datums/station_traits/neutral_traits.dm:227]: Would it be fine to use the `as anything` parameter here?

File: code/_onclick/hud/new_player.dm [hunk 4]
```diff
@@ -6,32 +6,53 @@
...
 			lobbyscreen.RegisterSignal(src, COMSIG_HUD_LOBBY_EXPANDED, TYPE_PROC_REF(/atom/movable/screen/lobby, expand_button))
-		if(istype(lobbyscreen, /atom/movable/screen/lobby/button))
+		if (istype(lobbyscreen, /atom/movable/screen/lobby/button))
 			var/atom/movable/screen/lobby/button/lobby_button = lobbyscreen
 			lobby_button.owner = REF(owner)
+	add_station_trait_buttons()
+
+/// Display buttons for relevant station traits
+/datum/hud/new_player/proc/add_station_trait_buttons()
+	if (!mymob || !mymob.client || mymob.client.interviewee || !length(GLOB.lobby_station_traits))
```
Comment (Ghommie) [inline code/_onclick/hud/new_player.dm:34]:
```suggestion
	if (!mymob?.client || mymob.client.interviewee || !length(GLOB.lobby_station_traits))
```
More or less like how you done it in the tooltip code.

**Review by Ghommie — APPROVED** [review]
I only have a couple comments on a few lines, but this is great. It gives me a few ideas to revamp the Ian and Pun Pun station traits too.

**Comment by projectkepler-ru** [comment]
This is goddam beautiful

**Review by Fikou — CHANGES_REQUESTED** [review]
mothblocks said "it should show on the normal job selection menu" (after round start) you havent added that

**Comment by Jacquerel** [comment]
> mothblocks said "it should show on the normal job selection menu" (after round start) you havent added that

oh yeah I missed that

**Comment by Jacquerel** [comment]
Changes:

Cargo Gorilla is now a job, the job handles most spawning.
Jobs can be set to only appear in the late join menu (we don't want you to set cargo gorilla employment preferences).
Cargo Gorilla appears in the latejoin menu if nobody signed up pre-round.

Cargo Gorilla (the job) can't be added to the available latejoin jobs by any player-accessible means... but it can by admins if they want to do that for some reason.

"Station trait which adds a new job" is now an abstract type ready for people to extend, there's not all that much code on the gorilla-specific station trait.

Because the Cargo Gorilla is now created via the interface of a job it just has its own random name list instead of copying whatever the sloth was called pre-deletion.

The cargo gorilla now gets a bank account, this is mostly because I tihnk it is funny.

**Comment by Mothblocks** [comment]
will look tmrw if i remember

File: code/modules/jobs/job_types/cargo_gorilla.dm [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	supervisors = SUPERVISOR_QM
+	spawn_type = /mob/living/basic/gorilla/cargorilla
+	config_tag = "CARGO_GORILLA"
+	random_spawns_possible = FALSE
+	display_order = JOB_DISPLAY_ORDER_CARGO_GORILLA
+	departments_list = list(/datum/job_department/cargo)
+	mail_goodies = list(
+		/obj/item/food/grown/banana = 1,
+	)
+	rpg_title = "Bananasmith"
```
Comment (Fikou) [inline code/modules/jobs/job_types/cargo_gorilla.dm:17]:
```suggestion
	rpg_title = "Merchant Mule"
```
they arent smiths and they dont have bananas

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Sorry, I had these pending for 5 days

File: code/game/gamemodes/game_mode.dm [hunk 6]
```diff
@@ -16,6 +16,8 @@
 
 ///Attempts to select players for special roles the mode might have.
 /datum/game_mode/proc/pre_setup()
+	SHOULD_CALL_PARENT(TRUE)
+	SEND_GLOBAL_SIGNAL(COMSIG_GLOB_PRE_GAMEMODE_SETUP, src)
 	return TRUE
```
[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/game/gamemodes/game_mode.dm:21]: It got merged

**Comment by Jacquerel** [comment]
dreamseeker_ntP3OayAuV
@Mothblocks I fixed it

**Comment by Jacquerel** [comment]
> Sorry, I had these pending for 5 days

also dont worry about that because i wouldnt have addressed them until today anyway lol

**Comment by Jacquerel** [comment]
> how does it look with multiple buttons now

image
But if it ever looks like this we fucked up

**Comment by Fikou** [comment]
why switch from a column to a row? it looked nicer in my version imo

**Comment by Jacquerel** [comment]
> why switch from a column to a row? it looked nicer in my version imo

ill be honest i didnt know it was originally a column, i can make it into one

**Comment by Jacquerel** [comment]
image
If it gets up to 4 then the fourth overlaps the menu in a weird looking way but i think if we have 4 traits with notifications active at once we have gone the wrong way (also you could add a second column or something)

I'll just codify that in the file I guess

**Pull request merged**